File: _56_PowerfulDigitSum.java
import java.util.*;
import java.math.BigInteger;

public class _56_PowerfulDigitSum{
    
    public static int getDigitSum(BigInteger bigInt){
        int sum = 0;
        String s = bigInt.toString();
        
        for (int i = 0; i < s.length(); i++){
            sum += s.charAt(i) - '0';
        }
        return sum;
    }
    
    public static void main(String[] args){
        
        int maxSum = 0;
        
        for (int a = 1; a < 100; a++){
            BigInteger bigInt = BigInteger.ONE;
            
            for (int b = 1; b < 100; b++){
                bigInt = bigInt.multiply(BigInteger.valueOf(a));
                int sum = getDigitSum(bigInt);
                // System.out.println(String.format("%d^%d = %s: %d", a, b, bigInt, sum));
                maxSum = sum > maxSum ? sum : maxSum;
            }
        }
        
        System.out.println("Maximum digital sum of a^b where a,b are natural numbers < 100: " + maxSum); // 972
    }
}

File: python/56_powerful_digit_sum.py
def powerful_digit_sum() -> int:
    
    max_sum: int = 0
    
    for a in range(1, 100):
        for b in range(1, 100):
            digit_sum = sum(int(chr) for chr in str(a ** b))
            max_sum = digit_sum if digit_sum > max_sum else max_sum
    return max_sum

if __name__ == "__main__":
    print(powerful_digit_sum())